### DISCUSSION TITLE
Feedback: Fast Data No Code Er Schema Configuration

### DISCUSSION THREAD

**Discussion by TheRusCompass** [mia-platform/community#227] [Product Feedback]
### Which section?

Fast Data

### Which is the version?

From v10.6.0

### Description

# Fast Data No Code Er Schema Configuration

With Console v10.6.0 there is the possibility to configure the Fast Data Er Schema in a No-Code fashion! What do you think about it?

Give us your feedback here!

## Feature Description

It is now possible to create and generate relationships between Projections using a simple and easy-to-use No Code configuration.
This tool is able to generate a JSON that will be used by the Single View Creator. The generated JSON is exactly the same JSON that would have been written manually in the Low Code Configuration.

In order to exploit this functionality you have to open the Projections section and navigate to the new ER Schema tab. You can drag&drop Projections inside a canvas and just connect them with a simple line! Then, just clicking on the link between two Projections, there will be the right panel where to configure the conditions and the relationships between them!

Once the Er Schema is completed, when configuring the Single View Creator, in the Single Views section, you just have to select one of the ER Schemas created in the aforementioned section from a dropdown list.

Check out the [official documentation](https://docs.mia-platform.eu/docs/fast_data/configuration/config_maps/erSchema#use_the_no_code) to know how to use it!

![er-schema-complete-view-d4c5f7154ac836e0df7fc2b2ab9e6deb](https://user-images.githubusercontent.com/117760936/230552900-62feea08-5ca6-4002-a7f8-46e0af1bbfc4.png)

## How to opt-in

The feature can be enabled by activating a Feature Toggle for a **specific Project** in the console for both PaaS and on-prem installation (in order to do that, you have to perform actions from the Console CMS). Reach out your Mia-Platform administrator in order to do that!

**Comment by flower-of-the-bridges** [comment]
After trying the new feature, I have the following suggestions:

### Old JSON ER Schema representation is misleading

When the No-Code feature is enabled, the Console automatically puts in the ER Schema section all the `erSchema.json` written previously and converts them from JSON documents to the ER visualization. However, their representation does not reflect the actual state of the condition defined in it.

For example, let's say that I have the following definition:

```
{
  "version": "1.0.0",
  "config": {
    "pr-a": {
      "outgoing": {
        "pr-b": {
          "conditions": {
            "pr-a_to_pr-b": {
              "condition": {
                "fieldB": "fieldA"
              }
            }
          }
        }
      }
    }
  }
}
```

This means that there is a 1:1, from `pr-a` to `pr-b`, but no vice-versa, i.e. I don't have stated specifically a way to reach `pr-a` from `pr-b`.
However, the erSchema shows that the condition is applied in both ways:

![image](https://user-images.githubusercontent.com/38134891/234818036-b1d9fe4b-436c-4d4e-9bd0-53315cce6a04.png)

### Support other Mongodb collections beside the one from the system of records

In some cases, ER Schemas may read from external collections of the Fast Data solution, such as CRUD collection or even other single views. 

This scenario can only be done by manually editing the JSON file generated by the view.

**Reply by TheRusCompass** [reply]
Thank you very much @giovf96 for this precious feedback! Let me explain some reasoning behind the implementation choices.

For what concern the first feedback (Old JSON ER Schema representation is misleading), the reason behind the described behavior is that we strongly recommend using both ways conditions even in the Low Code implementation, as you can see from the documentation (https://docs.mia-platform.eu/docs/fast_data/configuration/config_maps/erSchema#direction-of-the-relationships). That's why the no code feature is doing it for you. What we can take into consideration is to basically let the user choose if the declaration of the conditions in the opposite way is wanted or not.

For what concern the second feedback (Support other Mongodb collections beside the one from the system of records), this is a very interesting enhancement! We will for sure include it in our roadmap!

**Comment by nicola88** [comment]
Closing since we released Fast Data v2.
